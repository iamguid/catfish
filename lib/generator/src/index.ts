export * from './ProjectContext';
export * from './Plugin';
export * from './Project';
export * from './Templates';
export * from './plugins';