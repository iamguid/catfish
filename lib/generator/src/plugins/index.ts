export * as ProtobufModelsPlugin from './protobuf-models'