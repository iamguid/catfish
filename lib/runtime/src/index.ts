import * as pjs from "protobufjs/minimal";
import * as rxjs from "rxjs";
import grpc from "grpc-web";

export const convertRecordToMap = <TKey extends string | number, TInputVal, TOutputVal>(
    input: Record<TKey, TInputVal>,
    converter: (val: TInputVal) => TOutputVal
): Map<TKey, TOutputVal> => {
    const output = new Map();

    for (const prop in input) {
        output.set(prop, converter(input[prop]))
    }

    return output;
}

export const convertMapToRecord = <TKey extends string | number, TInputVal, TOutputVal>(
    input: Map<TKey, TInputVal>,
    converter: (val: TInputVal) => TOutputVal
): Record<TKey, TOutputVal> => {
    const output: Record<TKey, TOutputVal> = {} as Record<TKey, TOutputVal>;

    for (const [key, val] of input.entries()) {
        output[key] = converter(val);
    }

    return output;
}

export const convertBytesToBase64 = (bytes: Uint8Array | Buffer): string => {
    return pjs.util.base64.encode(bytes, 0, bytes.length);
}

export const convertBase64ToBytes = (base64: string): Uint8Array | Buffer => {
    const buffer: any = [];
    pjs.util.base64.decode(base64, buffer, 0);
    return Uint8Array.from(buffer);
}

export type ClientReadableStreamEvent<TData> = { type: "data", data: TData } | { type: 'metadata', metadata: grpc.Metadata } | { type: 'status', status: grpc.Status }

export const observableWrapClintReadableStream = <TResp>(stream: grpc.ClientReadableStream<TResp>): rxjs.Observable<ClientReadableStreamEvent<TResp>> => {
    return new rxjs.Observable((sbscriber) => {
        stream.on('data', (data) => {
            sbscriber.next({ type: 'data', data })
        })

        stream.on('error', (error) => {
            sbscriber.error(error)
        })

        stream.on('metadata', (metadata) => {
            sbscriber.next({ type: 'metadata', metadata })
        })

        stream.on('status', (status) => {
            sbscriber.next({ type: 'status', status })
        })

        stream.on('end', () => {
            sbscriber.complete();
        })

        return stream.cancel
    })
}