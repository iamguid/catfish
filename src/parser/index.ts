import { CharStreams, CommonTokenStream } from "antlr4";

import Protobuf3Lexer from "./generated/Protobuf3Lexer";
import Protobuf3Parser from "./generated/Protobuf3Parser";
import { FileDescriptor } from "./reflection/FileDescriptor";
import { Visitor } from "./visitor";

export * from "./reflection/index";

export const parse = (content: string): FileDescriptor => {
    const inputStream = CharStreams.fromString(content);
    const lexer = new Protobuf3Lexer(inputStream);
    const tokenStream = new CommonTokenStream(lexer);
    const parser = new Protobuf3Parser(tokenStream);
    const tree = parser.proto();
    const visitor = new Visitor();
    return visitor.visitProto(tree);
}
