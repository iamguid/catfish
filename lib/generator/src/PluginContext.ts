import { BaseDescriptor, EnumDescriptor, EnumFieldDescriptor, FileDescriptor, MapFieldDescriptor, MessageDescriptor, MessageFieldDescriptor, MethodDescriptor, OneofDescriptor, ServiceDescriptor } from "@catfish/parser";
import { Import, ProjectContext, TypeInfo } from "./ProjectContext";
import { ResolvedThing, ResolvedThingImport } from "./ResolverV2";

export type OptionalKey<T, key> = T extends Record<any, any> ? T[key] : unknown

export type PluginContextChainKeys = {
    readonly files: { I: any, O: any }
    readonly services: { I: any, O: any }
    readonly methods: { I: any, O: any }
    readonly messages: { I: any, O: any }
    readonly messagefields: { I: any, O: any }
    readonly maps: { I: any, O: any }
    readonly oneofs: { I: any, O: any }
    readonly enums: { I: any, O: any }
    readonly enumfields: { I: any, O: any }
    readonly typeinfos: { I: any, O: any }
}

export type BasePluginContext = {
    files: { I: FileDescriptor, O: { desc: FileDescriptor, filePath: string } }
    services: { I: ServiceDescriptor, O: { desc: ServiceDescriptor } }
    methods: { I: MethodDescriptor, O: { desc: MethodDescriptor } }
    messages: { I: MessageDescriptor, O: { desc: MessageDescriptor } }
    messagefields: { I: { msgDesc: MessageDescriptor, msgFieldDesc: MessageFieldDescriptor }, O: { msgDesc: MessageDescriptor, msgFieldDesc: MessageFieldDescriptor } }
    maps: { I: MapFieldDescriptor, O: { desc: MapFieldDescriptor } }
    oneofs: { I: OneofDescriptor, O: { desc: OneofDescriptor } }
    enums: { I: EnumDescriptor, O: { desc: EnumDescriptor } }
    enumfields: { I: { enmDesc: EnumDescriptor, enmFieldDesc: EnumFieldDescriptor }, O: { enmDesc: EnumDescriptor, enmFieldDesc: EnumFieldDescriptor } }
    typeinfos: { I: { type: string }, O: TypeInfo }
}

export type PluginContextResult<TPluginContext extends PluginContextChain<any, any>> = {
    [key in keyof TPluginContext]: OptionalKey<TPluginContext[key], 'O'>[]
}

export type PluginContextChain<TInput extends Partial<PluginContextChainKeys>, TOutput extends Partial<PluginContextChainKeys>> = {
    [key in keyof PluginContextChainKeys]: {
        I: TInput[key] extends Record<any, any> ? TInput[key]['I'] : never,
        O: (TInput[key] extends Record<any, any> ? TInput[key]['O'] : never)
            & (TOutput[key] extends Record<any, any> ? TOutput[key]['O'] : never)
    }
}

export type ContextsRegistryCb<
    TPluginOptions extends Record<string, any>,
    TPluginContext extends BasePluginContext = BasePluginContext,
    TName extends keyof TPluginContext = keyof TPluginContext,
> = (
    ctx: OptionalKey<TPluginContext[TName], 'I'>,
    prj: ProjectContext,
    file: FileDescriptor,
    use: (namespace: string, desc: BaseDescriptor, thingName: string) => ResolvedThingImport,
    def: (namespace: string, desc: BaseDescriptor) => Promise<ResolvedThing>,
    opt: TPluginOptions
) => Promise<OptionalKey<TPluginContext[TName], 'O'>>

export class ContextsRegistry<
    TPluginOptions extends Record<string, any>,
    TPluginContext extends BasePluginContext = BasePluginContext,
> {
    private registry: Record<string, ContextsRegistryCb<TPluginOptions, TPluginContext, keyof TPluginContext>[]> = {}

    constructor(
        private projectContext: ProjectContext,
        private file: FileDescriptor,
        private fileName: string,
        private opts: TPluginOptions,
    ) {
        this.buildBasePluginContext(this)
     }

    def = (namespace: string, desc: BaseDescriptor, thingName: string): ResolvedThing => {
        return this.projectContext.resolverV2.define(namespace, desc, thingName, this.fileName)
    }

    use = async (namespace: string, desc: BaseDescriptor, protoType?: string): Promise<ResolvedThingImport> => {
        return this.projectContext.resolverV2.resolveByNamespace(namespace, desc, this.file, protoType)
    }

    // extend = <
    //     TNextPluginContext,
    //     TName extends keyof TPluginContext
    // >(
    //     name: TName,
    //     cb: ContextsRegistryCb<TPluginOptions, TPluginContext, TName>
    // ): ContextsRegistry<PluginContextChain<TPluginContext, TNextPluginContext>> => {
    //     if (!this.registry[name as string]) {
    //         this.registry[name as string] = []
    //     }

    //     this.registry[name as string].push(cb)

    //     return this as ContextsRegistry<PluginContextChain<TPluginContext, TNextPluginContext>>
    // }

    extDesc <TName extends keyof TPluginContext>(
        name: TName,
        cb: ContextsRegistryCb<TPluginOptions, TPluginContext, TName>
    ) {
        if (!this.registry[name as string]) {
            this.registry[name as string] = []
        }

        this.registry[name as string].push(cb)

        return this;
    }


    build = <TResult = PluginContextResult<TPluginContext>, TF = TResult['files']>(): TResult => {
        const projectFiles = this.projectContext.getFiles();

        const files: TResult['files'] = [];
        const services: TResult['services'] = [];
        const methods: PluginContextResult<TPluginContext>['methods'] = [];
        const messages: PluginContextResult<TPluginContext>['messages'] = [];
        const messagefields: PluginContextResult<TPluginContext>['messagefields'] = [];
        const maps: PluginContextResult<TPluginContext>['maps'] = [];
        const oneofs: PluginContextResult<TPluginContext>['oneofs'] = [];
        const enums: PluginContextResult<TPluginContext>['enums'] = [];
        const enumfields: PluginContextResult<TPluginContext>['enumfields'] = [];
        const typeinfos: PluginContextResult<TPluginContext>['typeinfos'] = [];

        for (const prop in this.registry) {
            const cbChanin = this.registry[prop]

            // let lastResult = 
            // for (const cb of cbChanin) {
            //     cb()
            // }

        }

        return { files, services, methods, messages, messagefields, maps, oneofs, enums, enumfields, typeinfos }
    }

    private buildBasePluginContext = (cr: ContextsRegistry<BasePluginContext, Record<string, any>>) => {
        cr.extDesc('files', async (ctx, prj) => ({ desc: ctx, filePath: prj.getProtoFilePath(ctx) }))
        cr.extDesc('services', async (ctx) => ({ desc: ctx }))
        cr.extDesc('methods', async (ctx) => ({ desc: ctx }))
        cr.extDesc('messages', async (ctx) => ({ desc: ctx }))
        cr.extDesc('messagefields', async (ctx) => ({ msgDesc: ctx.msgDesc, msgFieldDesc: ctx.msgFieldDesc }))
        cr.extDesc('maps', async (ctx) => ({ desc: ctx }))
        cr.extDesc('oneofs', async (ctx) => ({ desc: ctx }))
        cr.extDesc('enums', async (ctx) => ({ desc: ctx }))
        cr.extDesc('enumfields', async (ctx) => ({ enmDesc: ctx.enmDesc, enmFieldDesc: ctx.enmFieldDesc }))
        cr.extDesc('typeinfos', async (ctx, prj, file) => prj.getTypeInfo(file, ctx.type))
    }
}

export type ExtendedPluignContext = PluginContextChain<BasePluginContext, {
    files: { I: BasePluginContext['files']['I'], O: { helloworld: string } }
}>

export const buildExtendedPluignContext = (cr: ContextsRegistry<Record<string, any>, ExtendedPluignContext>) => {
    cr.extDesc('files', async (ctx, prj) => ({ desc: ctx, filePath: prj.getProtoFilePath(ctx), helloworld: '' }))
}
