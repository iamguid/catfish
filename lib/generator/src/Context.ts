import fs from 'node:fs';
import path from 'node:path';

import { BaseDescriptor, FileDescriptor, parse } from "@catfish/parser"
import { walkByFiles } from "./fswalker"
import { ProjectOptions } from './Project';
import { wellKnownTypesToProtoFilesMap } from './wellKnownTypes';

export interface Import {
    name: string
    path: string
}

export interface TypeInfo {
    protoType: string
    descriptor?: BaseDescriptor
}

export interface Dependency {
    type: 'weak' | 'public'
    descriptor: FileDescriptor
}

export class Context {
    private readonly protoFiles = new Map<string, FileDescriptor>; // key - relative path to proto file
    private readonly filesPaths = new Map<FileDescriptor, string>; // value - relative path to proto file
    private readonly dependencies = new Map<FileDescriptor, Dependency[]>;
    
    constructor(private readonly config: ProjectOptions) {
    }

    // TODO: Make async
    load() {
        // Load protobuf well known types
        for (const [googlePath, realPath] of Object.entries(wellKnownTypesToProtoFilesMap)) {
            const parsed = parse(fs.readFileSync(realPath, 'utf8'));
            this.protoFiles.set(googlePath, parsed);
        }

        // Load project files
        walkByFiles({
            basePath: this.config.protoDirPath,
            filePath: "",
            each: ({ filePath }) => {
                if (path.extname(filePath) === '.proto') {
                    try {
                        const parsed = parse(fs.readFileSync(path.join(this.config.protoDirPath, filePath), 'utf8'));
                        this.protoFiles.set(filePath, parsed);
                    } catch (e) {
                        console.error(`Cannot parse file ${filePath}, ${(e as Error).message}`)
                    }
                }
            },
        });
    }

    // TODO: Make async
    resolve() {
        const resolveRecursive = (path: string, descriptor: FileDescriptor): Dependency[] => {
            const result: Dependency[] = [];

            for (const imprt of descriptor.imports) {
                if (!this.protoFiles.has(imprt.path)) {
                    throw new Error(`Cannot resolve import ${imprt.path} in file ${path}`)
                }
                
                const dependency = this.protoFiles.get(imprt.path)!;

                if (imprt.type === 'public') {
                    result.push({ type: 'public', descriptor: dependency });
                    result.push(...resolveRecursive(imprt.path, dependency));
                } else {
                    result.push({ type: 'weak', descriptor: dependency });
                }
            }

            return result;
        }

        for (const [path, descriptor] of this.protoFiles.entries()) {
            this.filesPaths.set(descriptor, path);

            if (!this.dependencies.has(descriptor)) {
                this.dependencies.set(descriptor, resolveRecursive(path, descriptor));
            }
        }
    }

    getFiles(includeWKT = false) {
        const files = Array.from(this.protoFiles.values());

        if (includeWKT) {
            return files;
        } else {
            return files.filter(f => {
                const filePath = this.getFilePathByDescriptor(f)
                return !(filePath in wellKnownTypesToProtoFilesMap)
            })
        }
    }

    getFilePathByDescriptor(descriptor: FileDescriptor) {
        if (!this.filesPaths.has(descriptor)) {
            throw new Error(`Cannot resolve file path by descriptor for file ${this.getFilePathByDescriptor(descriptor)}`)
        }

        return this.filesPaths.get(descriptor)!;
    }

    getDependencies(descriptor: FileDescriptor, includePublic = false) {
        if (!this.dependencies.has(descriptor)) {
            throw new Error(`There is no dependencies of file ${this.getFilePathByDescriptor(descriptor)}`)
        }

        const dependencies = this.dependencies.get(descriptor)!;

        if (includePublic) {
            return dependencies.map(d => d.descriptor);
        } else {
            return dependencies.filter(d => d.type !== 'public').map(d => d.descriptor)
        }
    }

    getProtoTypeDescriptor(descriptor: FileDescriptor, fullname: string) {
        const resolvedType = this.resolveType(descriptor, fullname);

        if (!resolvedType) {
            throw new Error(`There is no type descriptor ${fullname} for file ${this.getFilePathByDescriptor(descriptor)}`)
        }

        return resolvedType;
    }

    getTypeInfo(descriptor: FileDescriptor, protoType: string) {
        return this.extractTypeInfo(descriptor, protoType);
    }
    
    // TODO: Make cache
    private resolveType(fileDescriptor: FileDescriptor, fullname: string): BaseDescriptor | null {
        if (fileDescriptor.registry.descriptors.has(fullname)) {
            return fileDescriptor.registry.get(fullname);
        } else {
            for (const dependency of this.getDependencies(fileDescriptor, true)) {
                let resolvedType = dependency.registry.tryGet(fullname);

                if (resolvedType) {
                    return resolvedType
                }
            }
        }

        return null;
    }

    // TODO: make cache
    private extractTypeInfo(fileDescriptor: FileDescriptor, protoType: string): TypeInfo {
        const resolvedType = this.resolveType(fileDescriptor, protoType);

        return {
            protoType,
            descriptor: resolvedType ?? undefined
        }
    }
}