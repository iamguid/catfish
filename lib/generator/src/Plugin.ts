import { Context } from "./Context"
import { ProjectOptions } from "./Project"

export interface PluginOutputFile {
    path: string
    content: string
}

export interface PluginOutput {
    files: PluginOutputFile[]
}

export type Plugin<TOptions> = (
    context: Context,
    projectOptions: ProjectOptions,
    pluginOptions: TOptions
) => PluginOutput;
