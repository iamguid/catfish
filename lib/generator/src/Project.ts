import fs from 'node:fs';
import path from 'node:path';
import prettier from 'prettier';

import { ProjectContext } from './ProjectContext';
import { Plugin, PluginContextBuilder } from './Plugin';
import { Templates } from './Templates';

export interface ProjectOptions {
    protoDirPath: string
    outDirPath: string
}

export class Project {
    private context: ProjectContext;
    private plugins: [Plugin<any, any, any>, any, any][] = [];

    constructor(private readonly options: ProjectOptions) {
        this.context = new ProjectContext(options)
    }

    // TODO: make async
    load() {
        this.context.load();
        this.context.resolve();
        return this;
    }

    // TODO: make async
    generate() {
        fs.mkdirSync(this.options.outDirPath, { recursive: true })

        for (const [plugin, pluginContextBuilder, pluginOptions] of this.plugins) {
            const result = plugin(this.context, this.options, pluginOptions, pluginContextBuilder);

            for (const file of result.files) {
                const prettiedContent = prettier.format(file.content, { parser: 'typescript' })
                fs.writeFileSync(path.join(this.options.outDirPath, file.path), prettiedContent)
            }
        }
    }

    resgister<
        TContext,
        TTemplates extends Templates<TOptions, TTemplatesType>,
        TOptions extends Record<string, any>,
        TTemplatesType extends Record<string, (args: any[]) => string>
    >(
        plugin: Plugin<TContext, TOptions, TTemplates>,
        contextBuilder?: PluginContextBuilder<TContext, TOptions>,
        options?: TOptions,
    ) {
        this.plugins.push([plugin, contextBuilder, options]);
        return this;
    }
}