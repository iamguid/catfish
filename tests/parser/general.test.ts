import fs from "node:fs"
import { parse } from "../../src/parser"

describe("General proto parser test", () => {
    [
        "any_test.proto",
        "any.proto",
        "api.proto",
        "duration.proto",
        "empty.proto",
        "field_mask.proto",
        "map_proto3_unittest.proto",
        "map_unittest.proto",
        "struct.proto",
        "test_messages_proto3.proto",
        "timestamp.proto",
        "type.proto",
        "unittest_drop_unknown_fields.proto",
        "unittest_no_field_presence.proto",
        "unittest_preserve_unknown_enum.proto",
        "unittest_proto3_arena_lite.proto",
        "unittest_proto3_arena.proto",
        "unittest_proto3_lite.proto",
        "unittest_proto3_optional.proto",
        "unittest_proto3.proto",
        "unittest_well_known_types.proto",
        "wrappers.proto",
    ].forEach(protofile => {
        it(`Should parse ${protofile}`, async () => {
            const content = fs.readFileSync(`./tests/protos/${protofile}`);
            const tree = parse(content.toString());
            expect(tree.toObject()).toMatchSnapshot();
        })
    })
})