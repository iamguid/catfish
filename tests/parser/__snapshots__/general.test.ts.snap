// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`General proto parser test Should parse empty.proto 1`] = `
{
  "enums": [],
  "imports": [],
  "messages": [
    {
      "enums": [],
      "fields": [],
      "index": 0,
      "messages": [],
      "name": "Empty",
      "namespace": "",
      "options": [],
    },
  ],
  "options": [
    {
      "go_package": "google.golang.org/protobuf/types/known/emptypb",
    },
    {
      "java_package": "com.google.protobuf",
    },
    {
      "java_outer_classname": "EmptyProto",
    },
    {
      "java_multiple_files": true,
    },
    {
      "objc_class_prefix": "GPB",
    },
    {
      "csharp_namespace": "Google.Protobuf.WellKnownTypes",
    },
    {
      "cc_enable_arenas": true,
    },
  ],
  "package": "google.protobuf",
  "services": [],
}
`;
